connection_go18.go

```diff
@@ -0,0 +1,234 @@
+		return nil, err
+	}
+
+	var err error
+	var tx driver.Tx
+	if opts.ReadOnly {
+		tx, err = mc.beginReadOnly()
+	} else {
+		tx, err = mc.Begin()
+	}
+	mc.finish()
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		tx.Rollback()
+		return nil, ctx.Err()
+	}
+	return tx, err
+}
+
+func (mc *mysqlConn) beginReadOnly() (driver.Tx, error) {
+	if mc.isBroken() {
+		errLog.Print(ErrInvalidConn)
+		return nil, driver.ErrBadConn
+	}
+	// https://dev.mysql.com/doc/refman/5.7/en/innodb-performance-ro-txn.html
+	err := mc.exec("START TRANSACTION READ ONLY")
+	if err != nil {
+		return nil, err
+	}
+
+	return &mysqlTx{mc}, nil
+}
+
+func (mc *mysqlConn) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	rows, err := mc.Query(query, dargs)
+	if err != nil {
+		mc.finish()
+		return nil, err
+	}
+	if set, ok := rows.(setfinish); ok {
+		set.setFinish(mc.finish)
+	} else {
+		mc.finish()
+	}
+	return rows, err
+}
+
+func (mc *mysqlConn) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+	defer mc.finish()
+
+	return mc.Exec(query, dargs)
+}
+
+func (mc *mysqlConn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
+	if err := mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	stmt, err := mc.Prepare(query)
+	mc.finish()
+	if err != nil {
+		return nil, err
+	}
+
+	select {
+	default:
+	case <-ctx.Done():
+		stmt.Close()
+		return nil, ctx.Err()
+	}
+	return stmt, nil
+}
+
+func (stmt *mysqlStmt) QueryContext(ctx context.Context, args []driver.NamedValue) (driver.Rows, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := stmt.mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+
+	rows, err := stmt.Query(dargs)
+	if err != nil {
+		stmt.mc.finish()
+		return nil, err
+	}
+	if set, ok := rows.(setfinish); ok {
+		set.setFinish(stmt.mc.finish)
+	} else {
+		stmt.mc.finish()
+	}
+	return rows, err
+}
+
+func (stmt *mysqlStmt) ExecContext(ctx context.Context, args []driver.NamedValue) (driver.Result, error) {
+	dargs, err := namedValueToValue(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := stmt.mc.watchCancel(ctx); err != nil {
+		return nil, err
+	}
+	defer stmt.mc.finish()
+
+	return stmt.Exec(dargs)
+}
+
+func (mc *mysqlConn) watchCancel(ctx context.Context) error {
+	select {
+	default:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	if mc.watcher == nil {
+		return nil
+	}
+
+	mc.watcher <- ctx
+
+	return nil
+}
+
+func (mc *mysqlConn) startWatcher() {
+	watcher := make(chan mysqlContext, 1)
```

Comment:
there is watcher per connection, not one watcher for all connections, isn't it? 
Since mysql does not support multiplexing, only one query can be executed by the connection, so there is no problems IMHO.